TODO/making-photos-smaller.md

```diff
@@ -26,102 +26,102 @@ if format == 'JPEG':
 new_photo.save(thumbfile, **save_args)
 ```
 
-With this as a starting point, we began to investigate potential optimizations on file size that we could apply without a loss in quality.
+现在让我们开始研究研究有没有一种文件大小的优化方式可以以无损的形式进行。
 
-# Optimizations
+# 优化
 
-First, we had to decide whether to handle this ourselves or let a CDN provider [magically change](https://www.fastly.com/io) our photos. With the priority we place on high quality content, it made sense to evaluate options and make potential size vs quality tradeoffs ourselves. We moved ahead with research on the current state of photo file size reduction – what changes could be made and how much size / quality reduction was associated with each. With this research completed, we decided to work on three primary categories. The rest of this post explains what we did and how much benefit we realized from each optimization.
+首先，我们要决定是选择我们自己，还是一个 CDN 提供商 [magically change](https://www.fastly.com/io) 来处理我们的图片。在高质量内容优先的情况下，评估如何在图片大小和质量之间做取舍就显得非常重要了。让我们来研究一下当前图片文件减小的一些方法，我们可以做哪些改变以及每种方法我们可以减少多少大小和质量。完成这项研究之后，我们决定
 
-1. Changes in Pillow
-  - Optimize flag
-  - Progressive JPEG
-2. Changes to application photo logic
-  - Large PNG detection
-  - Dynamic JPEG quality
-3. Changes to JPEG encoder
-  - Mozjpeg (trellis quantization, custom quantization matrix)
+1. Pillow 中的改变
+- 优化 flag
+- 渐进式 JPEG
+2. 更改应用的照片逻辑
+- 大 PNG 检测
+- JPEG 动态质量
+3. 更换 JPEG 编码器
+- Mozjpeg (栅格量化，自定义量化矩阵)
 
-# Changes in Pillow
+# Pillow 中的改变
 
-## Optimize Flag
+## 优化 Flag
 
-This is one of the easiest changes we made: enabling the setting in Pillow responsible for additional file size savings at the cost of CPU time (`optimize=True`). Due to the nature of the tradeoff being made, this does not impact image quality at all.
+这是我们做的其中一个最简单的改变：启用 Pillow 中负责额外文件大小的设置来节约 CPU 耗时 (`optimize=True`)。由于本质没变，所有这对于图片质量丝毫没有影响。
 
-For JPEG, this flag instructs the encoder to find the optimal [Huffman coding](https://en.wikipedia.org/wiki/Huffman_coding) by making an additional pass over each image scan. Each first pass, instead of writing to file, calculates the occurrence statistics of each value, required information to compute the ideal coding. PNG internally uses zlib, so the optimize flag in that case effectively instructs the encoder to use `gzip -9` instead of `gzip -6`.
+对于 JPEG 来说，对个选项告诉编码器通过对每个图片进行一次额外的扫描以找到最佳的 [霍夫曼编码](https://en.wikipedia.org/wiki/Huffman_coding)。第一次，不写入文件，而是计算每个值出现的次数，以及可以计算出理想编码的必要信息。PNG 内部使用 zlib，所以优化 flag 就需要告诉编码器使用 `gzip -9` 而不是 `gzip -6`。
```

Comment:
PNG internally uses zlib, so the optimize flag in that case effectively instructs the encoder to use `gzip -9` instead of `gzip -6`.
PNG 内部使用 zlib，所以优化 flag 就需要告诉编码器使用 `gzip -9` 而不是 `gzip -6`。
->
PNG内部使用zlib,所以在此情况下这个优化选项实际上命令编译器使用 `gzip -9`而不是`gzip -6`。